### ISSUE TITLE
[Bug]: [benchmark][standalone][LRU] after upgrading image, the number of loaded segment is less then before

### ISSUE THREAD

**Issue by wangting0128** [milvus-io/milvus#32379]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: debug-chyezh-lru-dev_enable_asan-cbfb3cb-20240417
- Deployment mode(standalone or cluster): standalone
- MQ type(rocksmq, pulsar or kafka): pulsar   
- SDK version(e.g. pymilvus v2.0.0rc2): 2.4.0rc66
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

upgrade image argo task: fouramf-w2l7f-1
re-search argo task：lru-big-data-nmjqm

server:
```
NAME                                                              READY   STATUS                            RESTARTS        AGE     IP              NODE         NOMINATED NODE   READINESS GATES
lru-big-data-etcd-0                                               1/1     Running                           0               2d7h    10.104.17.116   4am-node23   <none>           <none>
lru-big-data-milvus-standalone-5b954b66cd-bwkxd                   1/1     Running                           0               51m     10.104.21.225   4am-node24   <none>           <none>
lru-big-data-minio-5bd4bc5cd6-6wfkx                               1/1     Running                           0               2d7h    10.104.21.125   4am-node24   <none>           <none>
lru-big-data-pulsar-bookie-0                                      1/1     Running                           0               2d7h    10.104.31.44    4am-node34   <none>           <none>
lru-big-data-pulsar-bookie-1                                      1/1     Running                           0               2d7h    10.104.30.211   4am-node38   <none>           <none>
lru-big-data-pulsar-bookie-2                                      1/1     Running                           0               2d7h    10.104.17.117   4am-node23   <none>           <none>
lru-big-data-pulsar-broker-0                                      1/1     Running                           0               2d7h    10.104.6.186    4am-node13   <none>           <none>
lru-big-data-pulsar-proxy-0                                       1/1     Running                           0               2d7h    10.104.6.185    4am-node13   <none>           <none>
lru-big-data-pulsar-recovery-0                                    1/1     Running                           0               2d7h    10.104.1.199    4am-node10   <none>           <none>
lru-big-data-pulsar-zookeeper-0                                   1/1     Running                           0               2d7h    10.104.17.115   4am-node23   <none>           <none>
lru-big-data-pulsar-zookeeper-1                                   1/1     Running                           0               2d7h    10.104.29.76    4am-node35   <none>           <none>
lru-big-data-pulsar-zookeeper-2                                   1/1     Running                           0               2d7h    10.104.34.168   4am-node37   <none>           <none>
```
<img width="1429" alt="img_v3_02a1_39c631c1-49e0-4cd6-af04-70e3fe37dfag" src="https://github.com/milvus-io/milvus/assets/26307815/3c5a12ce-7626-4d7a-9c16-c7889de4ca8c">

![img_v3_02a1_8e8adbf3-3aca-46f4-a2f5-fc969128128g](https://github.com/milvus-io/milvus/assets/26307815/bfad7572-4f7d-428d-bb5f-8d7894c71840)

<img width="1001" alt="img_v3_02a1_ddb5a7f9-9c34-4eb4-8e1e-3fe89ea7229g" src="https://github.com/milvus-io/milvus/assets/26307815/27474448-245f-4bc0-ac4c-eaa6ad8baeef">

![8RR6hkkCmX](https://github.com/milvus-io/milvus/assets/26307815/346582ac-d118-4087-8a5e-4a75acf629a2)

### Steps To Reproduce

```markdown
re-search collection after upgrading image
```

**Comment by chyezh** [comment]
There's a dead lock on querynode when releasing segment before upgrading.
Segment cannot be released after handoff. 
So, the segment on old cluster is greater than new cluster.

```
2024-04-17 15:46:09.077	
(no unique labels)
[2024/04/17 07:46:09.077 +00:00] [INFO] [task/executor.go:283] ["release segment..."] [taskID=1713334524048] [collectionID=449140365187052676] [replicaID=449140365344112641] [segmentID=449140365187653088] [node=11] [source=segment_checker] [shardLeader=11]
2024-04-17 15:46:09.077	
(no unique labels)
[2024/04/17 07:46:09.077 +00:00] [INFO] [querynodev2/services.go:528] ["received release segment request"] [traceID=e58983bc7e7d7cd4b8672a4069485037] [collectionID=449140365187052676] [shard=by-dev-rootcoord-dml_6_449140365187052676v0] [segmentIDs="[449140365187653088]"] [currentNodeID=11] [scope=Historical] [needTransfer=true]
2024-04-17 15:46:09.077	
(no unique labels)
[2024/04/17 07:46:09.077 +00:00] [DEBUG] [pipeline/filter_policy.go:114] ["add exclude info"] [segmentID=449140365187653088] [ts=18446744073709551615]
2024-04-17 15:46:09.077	
(no unique labels)
[2024/04/17 07:46:09.077 +00:00] [INFO] [delegator/delegator_data.go:790] ["delegator start to release segments"] [traceID=e58983bc7e7d7cd4b8672a4069485037] [collectionID=449140365187052676] [channel=by-dev-rootcoord-dml_6_449140365187052676v0] [replicaID=449140365344112641] [segmentIDs="[449140365187653088]"] [nodeID=11] [scope=Historical] [force=false]
2024-04-17 15:46:09.078	
(no unique labels)
[2024/04/17 07:46:09.078 +00:00] [INFO] [querynodev2/services.go:528] ["received release segment request"] [traceID=e58983bc7e7d7cd4b8672a4069485037] [collectionID=449140365187052676] [shard=by-dev-rootcoord-dml_6_449140365187052676v0] [segmentIDs="[449140365187653088]"] [currentNodeID=11] [scope=Historical] [needTransfer=false]
2024-04-17 15:46:09.078	
(no unique labels)
[2024/04/17 07:46:09.078 +00:00] [INFO] [querynodev2/services.go:568] ["start to release segments"] [traceID=e58983bc7e7d7cd4b8672a4069485037] [collectionID=449140365187052676] [shard=by-dev-rootcoord-dml_6_449140365187052676v0] [segmentIDs="[449140365187653088]"] [currentNodeID=11]
2024-04-17 15:46:09.577	
(no unique labels)
[2024/04/17 07:46:09.577 +00:00] [INFO] [task/scheduler.go:798] ["task removed"] [taskID=1713334524048] [collectionID=449140365187052676] [replicaID=449140365344112641] [status=succeeded] [segmentID=449140365187653088]
2024-04-17 15:46:09.578	
(no unique labels)
[2024/04/17 07:46:09.578 +00:00] [WARN] [task/executor.go:287] ["failed to release segment"] [taskID=1713334524048] [collectionID=449140365187052676] [replicaID=449140365344112641] [segmentID=449140365187653088] [node=11] [source=segment_checker] [shardLeader=11] [error="stack trace: /go/src/github.com/milvus-io/milvus/pkg/tracer/stack_trace.go:51 github.com/milvus-io/milvus/pkg/tracer.StackTrace\n/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:556 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).Call\n/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:570 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).ReCall\n/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:90 github.com/milvus-io/milvus/internal/distributed/querynode/client.wrapGrpcCall[...]\n/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:195 github.com/milvus-io/milvus/internal/distributed/querynode/client.(*Client).ReleaseSegments\n/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:164 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments.func1\n/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:271 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).send\n/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:161 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments\n/go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:284 github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).releaseSegment\n/go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:152 github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).executeSegmentAction: rpc error: code = Canceled desc = context canceled"] [errorVerbose="stack trace: /go/src/github.com/milvus-io/milvus/pkg/tracer/stack_trace.go:51 github.com/milvus-io/milvus/pkg/tracer.StackTrace: rpc error: code = Canceled desc = context canceled\n(1) attached stack trace\n  -- stack trace:\n  | github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).Call\n  | \t/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:556\n  | github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).ReCall\n  | \t/go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:570\n  | github.com/milvus-io/milvus/internal/distributed/querynode/client.wrapGrpcCall[...]\n  | \t/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:90\n  | github.com/milvus-io/milvus/internal/distributed/querynode/client.(*Client).ReleaseSegments\n  | \t/go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:195\n  | github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:164\n  | github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).send\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:271\n  | github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:161\n  | github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).releaseSegment\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:284\n  | github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).executeSegmentAction\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:152\n  | github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).Execute.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:121\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1598\nWraps: (2) stack trace: /go/src/github.com/milvus-io/milvus/pkg/tracer/stack_trace.go:51 github.com/milvus-io/milvus/pkg/tracer.StackTrace\n  | /go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:556 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).Call\n  | /go/src/github.com/milvus-io/milvus/internal/util/grpcclient/client.go:570 github.com/milvus-io/milvus/internal/util/grpcclient.(*ClientBase[...]).ReCall\n  | /go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:90 github.com/milvus-io/milvus/internal/distributed/querynode/client.wrapGrpcCall[...]\n  | /go/src/github.com/milvus-io/milvus/internal/distributed/querynode/client/client.go:195 github.com/milvus-io/milvus/internal/distributed/querynode/client.(*Client).ReleaseSegments\n  | /go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:164 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments.func1\n  | /go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:271 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).send\n  | /go/src/github.com/milvus-io/milvus/internal/querycoordv2/session/cluster.go:161 github.com/milvus-io/milvus/internal/querycoordv2/session.(*QueryCluster).ReleaseSegments\n  | /go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:284 github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).releaseSegment\n  | /go/src/github.com/milvus-io/milvus/internal/querycoordv2/task/executor.go:152 github.com/milvus-io/milvus/internal/querycoordv2/task.(*Executor).executeSegmentAction\nWraps: (3) rpc error: code = Canceled desc = context canceled\nError types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *status.Error"]
```

There's a search failure before release segment (already fixed by #32350)
Found that a reference count leak on streaming search.

**Comment by chyezh** [comment]
fixed by 2697c42e76ed0d587403d867d73df035b4765cbe,
@wangting0128  Please verify it.
